File: OOP/typescript/Polymorphism/Runtime.ts
class NavigationMode {
    navigate() {
        console.log("Calculating route...");
    }
}

class CarMode extends NavigationMode {
    navigate() {
        console.log("Car route selected: Fastest highways.");
    }
}

class BikeMode extends NavigationMode {
    navigate() {
        console.log("Bike route selected: Cycle-safe paths.");
    }
}

class WalkMode extends NavigationMode {
    navigate() {
        console.log("Walking route selected: Shortest footpath.");
    }
}

function startNavigation(mode: NavigationMode) {
    mode.navigate(); 
}

// Runtime decision
startNavigation(new CarMode());
startNavigation(new BikeMode());
startNavigation(new WalkMode());

File: OOP/java/Polymorphism/Runtime.java
// Parent class
package Polymorphism;
class NavigationMode {
    void navigate() {
        System.out.println("Calculating route...");
    }
}

// Child class 1
class CarMode extends NavigationMode {
    @Override
    void navigate() {
        System.out.println("Navigating by Car: Fastest road route selected.");
    }
}

// Child class 2
class BikeMode extends NavigationMode {
    @Override
    void navigate() {
        System.out.println("Navigating by Bike: Bike-friendly route selected.");
    }
}

// Child class 3
class WalkMode extends NavigationMode {
    @Override
    void navigate() {
        System.out.println("Navigating by Walking: Shortest footpath route selected.");
    }
}

public class Runtime {
    public static void main(String[] args) {

        NavigationMode mode;

        mode = new CarMode();
        mode.navigate();   // Car route

        mode = new BikeMode();
        mode.navigate();   // Bike route

        mode = new WalkMode();
        mode.navigate();   // Walk route
    }
}
